File: server/src/test/java/uk/co/unclealex/hammers/calendar/server/html/GameUpdateCommandTest.java
/**
 * Copyright 2011 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 *
 * @author unclealex72
 *
 */

package uk.co.unclealex.hammers.calendar.server.html;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

import junit.framework.Assert;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.junit.Test;

import uk.co.unclealex.hammers.calendar.server.html.GameUpdateCommand.Type;
import uk.co.unclealex.hammers.calendar.server.model.Game;
import uk.co.unclealex.hammers.calendar.server.model.GameKey;
import uk.co.unclealex.hammers.calendar.shared.model.Competition;
import uk.co.unclealex.hammers.calendar.shared.model.Location;

import com.google.common.base.Function;
import com.google.common.base.Joiner;
import com.google.common.base.Predicate;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;

/**
 * @author alex
 * 
 */
public class GameUpdateCommandTest {

	static final DateTime dateOf(int day, int month, int year) {
		return new DateTime(year, month, day, 15, 0, 0, 0, DateTimeZone.forID("Europe/London"));
	}

	private static final Competition DEFAULT_COMPETITION = Competition.FACP;
	private static final Location DEFAULT_LOCATION = Location.HOME;
	private static final String DEFAULT_OPPONENTS = "Them";
	private static final int DEFAULT_SEASON = 2012;
	private static final DateTime DEFAULT_DATE_PLAYED = dateOf(5, 9, 1972);
	private static final DateTime DEFAULT_BONDHOLDERS_AVAILABLE = dateOf(5, 9, 1973);
	private static final DateTime DEFAULT_PRIORITY_POINT_POST_AVAILABLE = dateOf(5, 9, 1974);
	private static final DateTime DEFAULT_SEASON_TICKETS_AVAILABLE = dateOf(5, 9, 1975);
	private static final DateTime DEFAULT_ACADEMY_TICKETS_AVAILABLE = dateOf(5, 9, 1976);
	private static final DateTime DEFAULT_GENERATE_SALE_TICKETS_AVAILABLE = dateOf(5, 9, 1977);
	private static final String DEFAULT_RESULT = "1-0";
	private static final Integer DEFAULT_ATTENDENCE = 100000;
	private static final String DEFAULT_MATCH_REPORT = "Good";
	private static final String DEFAULT_TELEVISION_CHANNEL = "BBC";
	private static final boolean DEFAULT_ATTENDED = false;

	/**
	 * Test there is a 1-1 mapping between orderings and game update command
	 * instances
	 */
	@Test
	public void testOrderingAndInstancesAreBijective() {
		Predicate<Method> isCreateGameUpdateCommandMethodPredicate = new Predicate<Method>() {
			@Override
			public boolean apply(Method method) {
				return GameUpdateCommand.class.equals(method.getReturnType());
			}
		};
		Function<Method, GameUpdateCommand> factory = new Function<Method, GameUpdateCommand>() {
			@Override
			public GameUpdateCommand apply(Method method) {
				try {
					return (GameUpdateCommand) method.invoke(null,
							GameLocator.gameKeyLocator(new GameKey(Competition.FACP, Location.HOME, "Them", 2012)), null);
				}
				catch (Throwable t) {
					Assert.fail("Could not invoke method " + method.getName());
					return null;
				}
			}
		};
		List<GameUpdateCommand> gameUpdateCommands = Lists
				.newArrayList(Iterables.transform(Iterables.filter(Arrays.asList(GameUpdateCommand.class.getMethods()),
						isCreateGameUpdateCommandMethodPredicate), factory));
		Function<GameUpdateCommand, GameUpdateCommand.Type> typeFunction = new Function<GameUpdateCommand, GameUpdateCommand.Type>() {
			@Override
			public Type apply(GameUpdateCommand gameUpdateCommand) {
				return gameUpdateCommand.getType();
			}
		};
		List<GameUpdateCommand.Type> missingTypes = Lists.newArrayList(GameUpdateCommand.Type.values());
		missingTypes.removeAll(Lists.newArrayList(Iterables.transform(gameUpdateCommands, typeFunction)));
		Assert.assertTrue("The following orderings are not covered: " + Joiner.on(", ").join(missingTypes),
				missingTypes.isEmpty());
		Assert.assertEquals("The wrong number of game update commands were found.", GameUpdateCommand.Type.values().length,
				gameUpdateCommands.size());
	}

	@Retention(RetentionPolicy.RUNTIME)
	@Target({ ElementType.METHOD })
	public @interface GameUpdateCommandTestCase {
	}

	@Test
	public void testAllGameUpdateCommandsAreTested() {
		Predicate<Method> isGameUpdateCommandTestCasePredicate = new Predicate<Method>() {
			@Override
			public boolean apply(Method method) {
				return method.getAnnotation(Test.class) != null
						&& method.getAnnotation(GameUpdateCommandTestCase.class) != null;
			}
		};
		Assert.assertEquals("The wrong number of game update command test cases were found.",
				GameUpdateCommand.Type.values().length,
				Iterables.size(Iterables.filter(Arrays.asList(getClass().getMethods()), isGameUpdateCommandTestCasePredicate)));
	}

	abstract class GameUpdateCommandFactory<E> {
		public abstract GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, E value);

		public GameUpdateCommand createGameUpdateCommand(GameKey gameKey, E value) {
			return createGameUpdateCommand(GameLocator.gameKeyLocator(gameKey), value);
		}
	}

	public <E> void testGameUpdateCommand(GameUpdateCommandFactory<E> gameUpdateCommandFactory,
			Function<Game, E> valueFunction, E currentValue, E newValue) {
		testNoChangeForNull(gameUpdateCommandFactory, valueFunction);
		testNullIsHandledCorrectly(gameUpdateCommandFactory, valueFunction, newValue);
		testNoChangeForEqualValue(gameUpdateCommandFactory, valueFunction, currentValue);
		testChangeForDifferentValues(gameUpdateCommandFactory, valueFunction, newValue);
	}

	protected <E> void testNullIsHandledCorrectly(GameUpdateCommandFactory<E> gameUpdateCommandFactory,
			Function<Game, E> valueFunction, E newValue) {
		Game game = new Game(1, DEFAULT_COMPETITION, DEFAULT_LOCATION, DEFAULT_OPPONENTS, DEFAULT_SEASON, null, null, null,
				null, null, null, null, null, null, null, false);
		GameUpdateCommand gameUpdateCommand = gameUpdateCommandFactory.createGameUpdateCommand(game.getGameKey(), newValue);
		Assert.assertTrue("A change was not made to a null value when one was expected.", gameUpdateCommand.update(game));
		Assert.assertNotNull("The changed value was null.", valueFunction.apply(game));
	}

	protected <E> void testNoChangeForNull(GameUpdateCommandFactory<E> gameUpdateCommandFactory,
			Function<Game, E> valueFunction) {
		Game game = createFullyPopulatedGame();
		GameUpdateCommand gameUpdateCommand = gameUpdateCommandFactory.createGameUpdateCommand(game.getGameKey(), null);
		Assert.assertFalse("A change was made for null when one was not expected.", gameUpdateCommand.update(game));
		Assert.assertNotNull("The changed value was null.", valueFunction.apply(game));
	}

	protected <E> void testNoChangeForEqualValue(GameUpdateCommandFactory<E> gameUpdateCommandFactory,
			Function<Game, E> valueFunction, E currentValue) {
		Game game = createFullyPopulatedGame();
		GameUpdateCommand gameUpdateCommand = gameUpdateCommandFactory.createGameUpdateCommand(game.getGameKey(),
				currentValue);
		Assert.assertFalse("A change was made for the current value when one was not expected.",
				gameUpdateCommand.update(game));
		Assert.assertEquals("The changed value was incorrect.", currentValue, valueFunction.apply(game));
	}

	protected <E> void testChangeForDifferentValues(GameUpdateCommandFactory<E> gameUpdateCommandFactory,
			Function<Game, E> valueFunction, E newValue) {
		Game game = createFullyPopulatedGame();
		GameUpdateCommand gameUpdateCommand = gameUpdateCommandFactory.createGameUpdateCommand(game.getGameKey(), newValue);
		Assert.assertTrue("A change was not made to the current value when one was expected.",
				gameUpdateCommand.update(game));
		Assert.assertEquals("The changed value was incorrect.", newValue, valueFunction.apply(game));
	}

	protected Game createFullyPopulatedGame() {
		return new Game(1, DEFAULT_COMPETITION, DEFAULT_LOCATION, DEFAULT_OPPONENTS, DEFAULT_SEASON, DEFAULT_DATE_PLAYED,
				DEFAULT_BONDHOLDERS_AVAILABLE, DEFAULT_PRIORITY_POINT_POST_AVAILABLE, DEFAULT_SEASON_TICKETS_AVAILABLE,
				DEFAULT_ACADEMY_TICKETS_AVAILABLE, DEFAULT_GENERATE_SALE_TICKETS_AVAILABLE, DEFAULT_RESULT, DEFAULT_ATTENDENCE,
				DEFAULT_MATCH_REPORT, DEFAULT_TELEVISION_CHANNEL, DEFAULT_ATTENDED);
	}

	@Test
	@GameUpdateCommandTestCase
	public void testDatePlayed() {
		GameUpdateCommandFactory<DateTime> gameUpdateCommandFactory = new GameUpdateCommandFactory<DateTime>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, DateTime newDatePlayed) {
				return GameUpdateCommand.datePlayed(gameLocator, newDatePlayed);
			}
		};
		Function<Game, DateTime> valueFunction = new Function<Game, DateTime>() {
			@Override
			public DateTime apply(Game game) {
				return game.getDatePlayed();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_DATE_PLAYED,
				DEFAULT_DATE_PLAYED.plusHours(1));
	}

	@Test
	@GameUpdateCommandTestCase
	public void testResult() {
		GameUpdateCommandFactory<String> gameUpdateCommandFactory = new GameUpdateCommandFactory<String>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, String newResult) {
				return GameUpdateCommand.result(gameLocator, newResult);
			}
		};
		Function<Game, String> valueFunction = new Function<Game, String>() {
			@Override
			public String apply(Game game) {
				return game.getResult();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_RESULT, "1" + DEFAULT_RESULT);
	}

	@Test
	@GameUpdateCommandTestCase
	public void testAttendence() {
		GameUpdateCommandFactory<Integer> gameUpdateCommandFactory = new GameUpdateCommandFactory<Integer>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, Integer newAttendence) {
				return GameUpdateCommand.attendence(gameLocator, newAttendence);
			}
		};
		Function<Game, Integer> valueFunction = new Function<Game, Integer>() {
			@Override
			public Integer apply(Game game) {
				return game.getAttendence();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_ATTENDENCE, DEFAULT_ATTENDENCE * 2);
	}

	@Test
	@GameUpdateCommandTestCase
	public void testMatchReport() {
		GameUpdateCommandFactory<String> gameUpdateCommandFactory = new GameUpdateCommandFactory<String>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, String newMatchReport) {
				return GameUpdateCommand.matchReport(gameLocator, newMatchReport);
			}
		};
		Function<Game, String> valueFunction = new Function<Game, String>() {
			@Override
			public String apply(Game game) {
				return game.getMatchReport();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_MATCH_REPORT, DEFAULT_MATCH_REPORT + "!");
	}

	@Test
	@GameUpdateCommandTestCase
	public void testTelevisionChannel() {
		GameUpdateCommandFactory<String> gameUpdateCommandFactory = new GameUpdateCommandFactory<String>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, String newTelevisionChannel) {
				return GameUpdateCommand.televisionChannel(gameLocator, newTelevisionChannel);
			}
		};
		Function<Game, String> valueFunction = new Function<Game, String>() {
			@Override
			public String apply(Game game) {
				return game.getTelevisionChannel();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_TELEVISION_CHANNEL,
				DEFAULT_TELEVISION_CHANNEL + "!");
	}

	@Test
	@GameUpdateCommandTestCase
	public void testAttended() {
		GameUpdateCommandFactory<Boolean> gameUpdateCommandFactory = new GameUpdateCommandFactory<Boolean>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, Boolean newAttended) {
				return GameUpdateCommand.attended(gameLocator, newAttended);
			}
		};
		Function<Game, Boolean> valueFunction = new Function<Game, Boolean>() {
			@Override
			public Boolean apply(Game game) {
				return game.isAttended();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_ATTENDED, !DEFAULT_ATTENDED);
	}

	@Test
	@GameUpdateCommandTestCase
	public void testBondHolder() {
		GameUpdateCommandFactory<DateTime> gameUpdateCommandFactory = new GameUpdateCommandFactory<DateTime>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, DateTime newDateTime) {
				return GameUpdateCommand.bondHolderTickets(gameLocator, newDateTime);
			}
		};
		Function<Game, DateTime> valueFunction = new Function<Game, DateTime>() {
			@Override
			public DateTime apply(Game game) {
				return game.getBondholdersAvailable();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_BONDHOLDERS_AVAILABLE,
				DEFAULT_BONDHOLDERS_AVAILABLE.plusHours(1));
	}

	@Test
	@GameUpdateCommandTestCase
	public void testPriorityTickets() {
		GameUpdateCommandFactory<DateTime> gameUpdateCommandFactory = new GameUpdateCommandFactory<DateTime>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, DateTime newDateTime) {
				return GameUpdateCommand.priorityPointTickets(gameLocator, newDateTime);
			}
		};
		Function<Game, DateTime> valueFunction = new Function<Game, DateTime>() {
			@Override
			public DateTime apply(Game game) {
				return game.getPriorityPointPostAvailable();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_PRIORITY_POINT_POST_AVAILABLE,
				DEFAULT_PRIORITY_POINT_POST_AVAILABLE.plusHours(1));
	}

	@Test
	@GameUpdateCommandTestCase
	public void testSeasonTickets() {
		GameUpdateCommandFactory<DateTime> gameUpdateCommandFactory = new GameUpdateCommandFactory<DateTime>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, DateTime newDateTime) {
				return GameUpdateCommand.seasonTickets(gameLocator, newDateTime);
			}
		};
		Function<Game, DateTime> valueFunction = new Function<Game, DateTime>() {
			@Override
			public DateTime apply(Game game) {
				return game.getSeasonTicketsAvailable();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_SEASON_TICKETS_AVAILABLE,
				DEFAULT_SEASON_TICKETS_AVAILABLE.plusHours(1));
	}

	@Test
	@GameUpdateCommandTestCase
	public void testAcademyTickets() {
		GameUpdateCommandFactory<DateTime> gameUpdateCommandFactory = new GameUpdateCommandFactory<DateTime>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, DateTime newDateTime) {
				return GameUpdateCommand.academyTickets(gameLocator, newDateTime);
			}
		};
		Function<Game, DateTime> valueFunction = new Function<Game, DateTime>() {
			@Override
			public DateTime apply(Game game) {
				return game.getAcademyMembersAvailable();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_ACADEMY_TICKETS_AVAILABLE,
				DEFAULT_ACADEMY_TICKETS_AVAILABLE.plusHours(1));
	}

	@Test
	@GameUpdateCommandTestCase
	public void testGeneralSaleTickets() {
		GameUpdateCommandFactory<DateTime> gameUpdateCommandFactory = new GameUpdateCommandFactory<DateTime>() {
			@Override
			public GameUpdateCommand createGameUpdateCommand(GameLocator gameLocator, DateTime newDateTime) {
				return GameUpdateCommand.generalSaleTickets(gameLocator, newDateTime);
			}
		};
		Function<Game, DateTime> valueFunction = new Function<Game, DateTime>() {
			@Override
			public DateTime apply(Game game) {
				return game.getGeneralSaleAvailable();
			}
		};
		testGameUpdateCommand(gameUpdateCommandFactory, valueFunction, DEFAULT_GENERATE_SALE_TICKETS_AVAILABLE,
				DEFAULT_GENERATE_SALE_TICKETS_AVAILABLE.plusHours(1));
	}
}

File: src/test/scala/uk/co/unclealex/hammers/calendar/html/GameUpdateCommandTest.scala
/**
 * Copyright 2010-2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

package uk.co.unclealex.hammers.calendar.html;

import java.lang.annotation.ElementType
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy
import java.lang.annotation.Target
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import uk.co.unclealex.hammers.calendar.model.Game
import uk.co.unclealex.hammers.calendar.model.GameKey
import uk.co.unclealex.hammers.calendar.model.Competition
import uk.co.unclealex.hammers.calendar.model.Location
import org.specs2.mutable.Specification
import uk.co.unclealex.hammers.calendar.Date._
import uk.co.unclealex.hammers.calendar.September
import scala.reflect.ClassTag

/**
 * The Class GameUpdateCommandTest.
 *
 * @author alex
 */
class GameUpdateCommandTest extends Specification {

  val DEFAULT_COMPETITION = Competition.FACP
  val DEFAULT_LOCATION = Location.HOME
  val DEFAULT_OPPONENTS = "Them"
  val DEFAULT_SEASON = 2012
  val DEFAULT_DATE_PLAYED = September(5, 1972)
  val DEFAULT_BONDHOLDERS_AVAILABLE = September(5, 1973)
  val DEFAULT_PRIORITY_POINT_POST_AVAILABLE = September(5, 1974)
  val DEFAULT_SEASON_TICKETS_AVAILABLE = September(5, 1975)
  val DEFAULT_ACADEMY_TICKETS_AVAILABLE = September(5, 1976)
  val DEFAULT_GENERAL_SALE_TICKETS_AVAILABLE = September(5, 1977)
  val DEFAULT_RESULT = "1-0"
  val DEFAULT_ATTENDENCE = 100000
  val DEFAULT_MATCH_REPORT = "Good"
  val DEFAULT_TELEVISION_CHANNEL = "BBC"
  val DEFAULT_ATTENDED = false

  /**
   * Test date played.
   */
  "Updating the date played" should {
    testGameUpdateCommand[DateTime](
      gameLocator => (datePlayed: Option[DateTime]) => DatePlayedUpdateCommand(gameLocator, datePlayed),
      game => game.dateTimePlayed,
      DEFAULT_DATE_PLAYED,
      DEFAULT_DATE_PLAYED plusHours 1)
  }

  "Updating the result" should {
    testGameUpdateCommand(
      gameLocator => (result: Option[String]) => ResultUpdateCommand(gameLocator, result),
      game => game.result,
      DEFAULT_RESULT,
      "1" + DEFAULT_RESULT)
  }

  "Updating the attendence" should {
    testGameUpdateCommand(
      gameLocator => (attendence: Option[Int]) => AttendenceUpdateCommand(gameLocator, attendence),
      game => game.attendence,
      DEFAULT_ATTENDENCE,
      DEFAULT_ATTENDENCE * 2)
  }

  "Updating the match report" should {
    testGameUpdateCommand(
      gameLocator => (matchReport: Option[String]) => MatchReportUpdateCommand(gameLocator, matchReport),
      game => game.matchReport,
      DEFAULT_MATCH_REPORT,
      DEFAULT_MATCH_REPORT + "!")
  }

  "Updating the television channel" should {
    testGameUpdateCommand(
      gameLocator => (televisionChannel: Option[String]) => TelevisionChannelUpdateCommand(gameLocator, televisionChannel),
      game => game.televisionChannel,
      DEFAULT_TELEVISION_CHANNEL,
      DEFAULT_TELEVISION_CHANNEL + "!")
  }

  "Updating the attended flag" should {
    testGameUpdateCommand(
      gameLocator => (attended: Option[Boolean]) => AttendedUpdateCommand(gameLocator, attended),
      game => game.attended,
      DEFAULT_ATTENDED,
      !DEFAULT_ATTENDED)
  }

  "Updating the bond holder ticket sale date" should {
    testGameUpdateCommand[DateTime](
      gameLocator => (saleDate: Option[DateTime]) => BondHolderTicketsUpdateCommand(gameLocator, saleDate),
      game => game.dateTimeBondholdersAvailable,
      DEFAULT_BONDHOLDERS_AVAILABLE,
      DEFAULT_BONDHOLDERS_AVAILABLE plusDays 1)
  }

  "Updating the priority point ticket sale date" should {
    testGameUpdateCommand[DateTime](
      gameLocator => (saleDate: Option[DateTime]) => PriorityPointTicketsUpdateCommand(gameLocator, saleDate),
      game => game.dateTimePriorityPointPostAvailable,
      DEFAULT_PRIORITY_POINT_POST_AVAILABLE,
      DEFAULT_PRIORITY_POINT_POST_AVAILABLE plusDays 1)
  }

  "Updating the season ticket holders' ticket sale date" should {
    testGameUpdateCommand[DateTime](
      gameLocator => (saleDate: Option[DateTime]) => SeasonTicketsUpdateCommand(gameLocator, saleDate),
      game => game.dateTimeSeasonTicketsAvailable,
      DEFAULT_SEASON_TICKETS_AVAILABLE,
      DEFAULT_SEASON_TICKETS_AVAILABLE plusDays 1)
  }

  "Updating the academy members' ticket sale date" should {
    testGameUpdateCommand[DateTime](
      gameLocator => (saleDate: Option[DateTime]) => AcademyTicketsUpdateCommand(gameLocator, saleDate),
      game => game.dateTimeAcademyMembersAvailable,
      DEFAULT_ACADEMY_TICKETS_AVAILABLE,
      DEFAULT_ACADEMY_TICKETS_AVAILABLE plusDays 1)
  }

  "Updating the general ticket sale date" should {
    testGameUpdateCommand[DateTime](
      gameLocator => (saleDate: Option[DateTime]) => GeneralSaleTicketsUpdateCommand(gameLocator, saleDate),
      game => game.dateTimeGeneralSaleAvailable,
      DEFAULT_GENERAL_SALE_TICKETS_AVAILABLE,
      DEFAULT_GENERAL_SALE_TICKETS_AVAILABLE plusDays 1)
  }

  /**
   * Test a game update command.
   *
   * @param gameUpdateCommandFactory
   *          the game update command factory
   * @param valueFunction
   *          the value function
   * @param currentValue
   *          the current value
   * @param newValue
   *          the new value
   */
  def testGameUpdateCommand[E](
    gameUpdateCommandFactory: GameLocator => Option[E] => GameUpdateCommand,
    valueFactory: Game => Option[E],
    currentValue: E,
    newValue: E)(implicit ct: ClassTag[E]) {
    val updateCommandFactory = (game: Game) => gameUpdateCommandFactory(GameKeyLocator(game.gameKey))
    "not change for None values" in {
      testNoChangeForNone(updateCommandFactory, valueFactory)
    }
    "not change for equal values" in {
      testNoChangeForEqualValue(updateCommandFactory, valueFactory, currentValue)
    }
    "change for different values" in {
      testChangeForDifferentValues(updateCommandFactory, valueFactory, newValue)
    }
  }

  /**
   * Test no change for null.
   *
   * @param <E>
   *          the element type
   * @param gameUpdateCommandFactory
   *          the game update command factory
   * @param valueFunction
   *          the value function
   */
  def testNoChangeForNone[E](
    gameUpdateCommandFactory: Game => Option[E] => GameUpdateCommand,
    valueFactory: Game => Option[E]) {
    val game = createFullyPopulatedGame
    val gameUpdateCommand = gameUpdateCommandFactory(game)(None: Option[E])
    gameUpdateCommand update game must be equalTo (false)
    valueFactory(game) must not be equalTo(None)
  }

  /**
   * Test no change for equal value.
   *
   * @param <E>
   *          the element type
   * @param gameUpdateCommandFactory
   *          the game update command factory
   * @param valueFunction
   *          the value function
   * @param currentValue
   *          the current value
   */
  def testNoChangeForEqualValue[E](
    gameUpdateCommandFactory: Game => Option[E] => GameUpdateCommand,
    valueFactory: Game => Option[E],
    currentValue: E) {
    val game = createFullyPopulatedGame
    val gameUpdateCommand = gameUpdateCommandFactory(game)(Some(currentValue))
    gameUpdateCommand update game should be equalTo (false)
    valueFactory(game) should be equalTo (Some(currentValue))
  }

  /**
   * Test change for different values.
   *
   * @param <E>
   *          the element type
   * @param gameUpdateCommandFactory
   *          the game update command factory
   * @param valueFunction
   *          the value function
   * @param newValue
   *          the new value
   */
  def testChangeForDifferentValues[E](
    gameUpdateCommandFactory: Game => Option[E] => GameUpdateCommand,
    valueFactory: Game => Option[E],
    newValue: E) {
    val game = createFullyPopulatedGame
    val gameUpdateCommand =
      gameUpdateCommandFactory(game)(Some(newValue))
    gameUpdateCommand update game should be equalTo (true)
    valueFactory.apply(game) should be equalTo (Some(newValue))
  }

  /**
   * Creates the fully populated game.
   *
   * @return the game
   */
  def createFullyPopulatedGame: Game = {
    return new Game(
      1,
      DEFAULT_COMPETITION,
      DEFAULT_LOCATION,
      DEFAULT_OPPONENTS,
      DEFAULT_SEASON,
      Some(DEFAULT_DATE_PLAYED),
      Some(DEFAULT_BONDHOLDERS_AVAILABLE),
      Some(DEFAULT_PRIORITY_POINT_POST_AVAILABLE),
      Some(DEFAULT_SEASON_TICKETS_AVAILABLE),
      Some(DEFAULT_ACADEMY_TICKETS_AVAILABLE),
      Some(DEFAULT_GENERAL_SALE_TICKETS_AVAILABLE),
      Some(DEFAULT_RESULT),
      Some(DEFAULT_ATTENDENCE),
      Some(DEFAULT_MATCH_REPORT),
      Some(DEFAULT_TELEVISION_CHANNEL),
      Some(DEFAULT_ATTENDED))
  }

}
